Wrappers/Python/test/test_io.py

```diff
@@ -150,3 +155,96 @@ def test_read_and_reconstruct_2D(self):
+    def setUp(self) -> None:
+        self.logger = logging.getLogger('cil.io')
+        self.logger.setLevel(logging.DEBUG)
+        self.cwd = os.path.join(os.getcwd(), 'tifftest')
+        # print("creating {}".format(self.cwd))
+        os.mkdir(self.cwd)
+    def tearDown(self) -> None:
+        # print("removing {}".format(self.cwd))
+        shutil.rmtree(self.cwd)
+        # print("done")
+
+    def test_tiff_stack_ImageData(self):
+        data = dataexample.SIMULATED_SPHERE_VOLUME.get()
+        
+        fname = os.path.join(self.cwd, "unittest")
+
+        writer = TIFFWriter(data=data, file_name=fname)
+        writer.write()
+
+        reader = TIFFStackReader(file_name=self.cwd)
+        read_array = reader.read()
+
+        np.testing.assert_allclose(data.as_array(), read_array)
+
+        read = reader.read_as_ImageData(data.geometry)
+        np.testing.assert_allclose(data.as_array(), read.as_array())
+
+    def test_tiff_stack_AcquisitionData(self):
+        data = dataexample.SIMULATED_CONE_BEAM_DATA.get()
+        
+        fname = os.path.join(self.cwd, "unittest")
+
+        writer = TIFFWriter(data=data, file_name=fname)
+        writer.write()
+
+        reader = TIFFStackReader(file_name=self.cwd)
+        read_array = reader.read()
+
+        np.testing.assert_allclose(data.as_array(), read_array)
+
+        read = reader.read_as_AcquisitionData(data.geometry)
+        np.testing.assert_allclose(data.as_array(), read.as_array())
+
+    def test_tiff_stack_ImageDataSlice(self):
+        data = dataexample.SIMULATED_SPHERE_VOLUME.get()
+        
+        fname = os.path.join(self.cwd, "unittest")
+
+        writer = TIFFWriter(data=data, file_name=fname)
+        writer.write()
+
+        roi = {'axis_0': -1, 'axis_1': -1, 'axis_2': (None, None, 2)}
+
+        reader = TIFFStackReader(file_name=self.cwd, roi=roi, mode='slice')
+        read_array = reader.read()
+
+        shape = [el for el in data.shape]
+        shape[2] /= 2
+
+        np.testing.assert_allclose(shape, read_array.shape )
+        
+        roi = {'axis_0': (0, 2, None), 'axis_1': -1, 'axis_2': -1}
+
+        reader = TIFFStackReader(file_name=self.cwd, roi=roi, mode='slice')
+        read_array = reader.read()
+
+        np.testing.assert_allclose(data.as_array()[:2], read_array)
+
+    def test_tiff_stack_ImageData_wrong_file(self):
+        data = dataexample.SIMULATED_SPHERE_VOLUME.get()
+        
+        fname = os.path.join(self.cwd, "unittest")
+
+        writer = TIFFWriter(data=data, file_name=fname)
+        writer.write()
+
+        import glob
+        for el in glob.glob(os.path.join(self.cwd , "unittest*.tiff")):
+            # print (f"modifying {el}")
+            with open(el, 'w') as f:
+                f.write('BOOM')
+            break
+                
+        reader = TIFFStackReader(file_name=self.cwd)
+        try:
+            read_array = reader.read()
+            assert False, 'WHAT???? Files should not load'
```

Comment:
:D Not that I don't love the comments... but we should only have output on a unit test failure

Reply:
yes, it it reaches that `assert` has failed